src/lib/stream/shake_cipher/shake_cipher.h

```diff
@@ -36,20 +37,47 @@ class SHAKE_128_Cipher final : public StreamCipher
       */
       void set_iv(const uint8_t iv[], size_t iv_len) override;
 
-      Key_Length_Specification key_spec() const override;
-
       void clear() override;
```

Comment:
Assuming that `SHAKE_Cipher` won't be in the public interface, I'd argue that `SHAKE_xxx_Cipher` are marked final as a whole instead.

Reply:
Any class that can be marked `final` definitely should since that helps the optimizer. I like using `final` annotations on specific functions where they are implemented on some class that is not itself `final`, but where we expect/know that the subclasses will not override that function. I don't think this usage helps the optimizer much but it is imo useful from a documentation/code understanding perspective. We promise not to override the function further, *and* the compiler checks that we keep our promise.